### DISCUSSION TITLE
What if there are tremendous names in ZAP having the exact same hash?

### DISCUSSION THREAD

**Discussion by outofforest** [openzfs/zfs#15191] [General]
Hello everyone,

I know that situation I'm asking about is almost impossible and probably it will never happen before the end of the world but I'm still interested in getting an answer to this theoretical question.

Let's say that I have tremendous (going to infinity) number of names to store in ZAP and hashes of all those names are identical. Then all those names go to the same leaf node in the ZAP tree. Obviously the available space in that node is exhausted. Splitting the data, by adding another pointer node won't solve the problem because after redistribution all the records will go to the same node again.

Is there any way of handling this situation implemented in ZAP? Or you just assume that nobody ever is so unlucky?

-- 
Thanks!
Wojciech

**Comment by amotin** [comment]
Its a good question what would happen if one block is not enough to store it.  Original ZFS design allowed block chaining there, but it was never implemented.  Though chance of such collision should be pretty low.  And IIRC there is a random salt added to the hash make attack on this difficult.
